### PR TITLE
Added Utility Types in Typescript

### LINKED ISSUE #5091: [Entry] Subject: Entry Add Utility Types to Typescript
### Reason/inspiration (optional)

Add the utility types to the typescript section.

### Entry Type (select all that apply)

New Concept Entry for Existing Topic

### Code of Conduct

- [X] By submitting this issue, I agree to follow Codecademy Doc's [Code of Conduct](https://github.com/Codecademy/docs/blob/main/.github/CODE_OF_CONDUCT.md).

### For Maintainers

- [x] Labels added
- [x] Issue is assigned

### PR DESCRIPTION
**Pull request by Mandipbhattarai**
### Issue Solved
  Fixes #5091

### Type of Change

- Adding a new entry
- Editing an existing entry (fixing a typo, bug, issues, etc)
- Updating the documentation

### Checklist

- [x] All writings are my own.
- [x] My entry follows the Codecademy Docs style guide.
- [x] My changes generate no new warnings.
- [x] I have performed a self-review of my own writing and code.
- [x] I have checked my entry and corrected any misspellings.
- [x] I have made corresponding changes to the documentation if needed.
- [x] I have confirmed my changes are not being pushed from my forked `main` branch.
- [x] I have confirmed that I'm pushing from a new branch named after the changes I'm making.
- [x] I have linked any issues that are relevant to this PR in the `Issues Solved` section.

### CODE REVIEW

**Comment by Mandipbhattarai** [comment]
Sure working on it. Thanks for the review!

**Comment by Mandipbhattarai** [comment]
I have applied the changes. Can you review it @dakshdeepHERE?

**Comment by Mandipbhattarai** [comment]
@dakshdeepHERE is there any other changes I must apply?

**Comment by Mandipbhattarai** [comment]
Hi @dakshdeepHERE ! Can you please provide me with the update?

**Comment by Mandipbhattarai** [comment]
@SaviDahegaonkar why is the review taking so long time?

**Comment by Mandipbhattarai** [comment]
@SaviDahegaonkar thanks for the response!

File: content/typescript/concepts/utility-types/utility-types.md [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+Tags:
+  - 'Utility Types'
+  - 'Data Types'
+  - 'TypeScript'
+CatalogContent:
+  - 'learn-typescript'
+  - 'paths/full-stack-engineer-career-path'
+---
+
+TypeScript consists of several predefined generic types known as Utility Types. These utility types are useful for manipulating or creating new types more efficiently.
```
Comment (SaviDahegaonkar) [inline content/typescript/concepts/utility-types/utility-types.md:17]:
```suggestion
TypeScript consists of several predefined generic types known as **Utility Types**. These utility types are useful for manipulating or creating new types more efficiently.
```
The first mention of the concept must be in bold.

File: content/typescript/concepts/utility-types/utility-types.md [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+
+## Partial
+
+Partial is used to create a type with all properties marked as optional.
```
Comment (SaviDahegaonkar) [inline content/typescript/concepts/utility-types/utility-types.md:21]:
```suggestion
Partial is  a utility type that is used to create a type where all properties are marked as optional.
```

File: content/typescript/concepts/utility-types/utility-types.md [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+}
+
+const partialUser: Partial<User> = {
+  name: 'John',
+};
+```
+
+## Required
+
+`Required` is used to create a type with all properties of `T` marked as `Required`. It is the opposite of `Partial`.
```
Comment (SaviDahegaonkar) [inline content/typescript/concepts/utility-types/utility-types.md:51]:
```suggestion
`Required` is a utility type that is used to create a type with all properties of `T` marked as `Required`. It is the opposite of `Partial`.
```

File: content/typescript/concepts/utility-types/utility-types.md [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+
+In this example, all the properties of the `User` interface (including the optional properties) are made mandatory:
+
+```ts
+interface User {
+  name: string;
+  age?: number;
+}
+
+const requiredUser: Required<User> = {
```
Comment (SaviDahegaonkar) [inline content/typescript/concepts/utility-types/utility-types.md:73]:
```suggestion
const requiredUser: Required<{ name: string; age: number}> = {
```
When we use Required utility type we must make sure to provide the values for all the properties that we have defined.

File: content/typescript/concepts/utility-types/utility-types.md [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+  name: 'John',
+  age: 25,
+};
+```
+
+## Readonly
+
+`Readonly` is to mark all the properties in `T` as `Readonly`, which prevents any reassignment.
```
Comment (SaviDahegaonkar) [inline content/typescript/concepts/utility-types/utility-types.md:81]:
```suggestion
`Readonly`  is a utility type that marks all the properties in `T` as `readonly`, preventing any reassignment.
```

File: content/typescript/concepts/utility-types/utility-types.md [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+
+### Syntax
+
+Exclude is written as:
+
+```pseudo
+Exclude<T, U>;
+```
+
+where `U` are the properties you want to exclude from `T`.
```
Comment (SaviDahegaonkar) [inline content/typescript/concepts/utility-types/utility-types.md:213]:
```suggestion
where `U` represents the types you want to exclude from the union type`T`.
```

File: content/typescript/concepts/utility-types/utility-types.md [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+
+### Exclude Example
+
+In this example, we exclude `apple` and `banana` from the union type, resulting in just `orange`:
+
+```ts
+type fruits = 'apple' | 'banana' | 'orange';
+type ExcludedFruits = Exclude<T, 'apple' | 'banana'>; // "orange"
```
Comment (SaviDahegaonkar) [inline content/typescript/concepts/utility-types/utility-types.md:221]:
```suggestion
type ExcludedFruits = Exclude<fruits, 'apple' | 'banana'>; // "orange"
```

**Review by SaviDahegaonkar — CHANGES_REQUESTED** [review]
Hey @Mandipbhattarai 
I have suggested a few changes please make them and let me know. Also I want you to include the stage 'Utility types' in the [tags.md file](documentation/tags.md).

**Comment by Mandipbhattarai** [comment]
@SaviDahegaonkar as per your changes i have updated it and also included the utility types.

**Comment by Mandipbhattarai** [comment]
Can you recheck it @SaviDahegaonkar because I had already added the tag.
image

**Comment by SaviDahegaonkar** [comment]
Hey @Mandipbhattarai , Are you sure you added the tag in the tags.md list for the forked repository because as you can see I can't find any 'utility types' tag . 
Screenshot 2024-10-15 165524

**Comment by Mandipbhattarai** [comment]
@SaviDahegaonkar can you check the file changed because it shows the addition of 'Utility Types' in the tags.md from my pr.
image

**Comment by SaviDahegaonkar** [comment]
> @SaviDahegaonkar can you check the file changed because it shows the addition of 'Utility Types' in the tags.md from my pr. image

Yeah I got it, it was a network issue from my end so I was unable to find the reflected changes.

**Review by SaviDahegaonkar — APPROVED** [review]
Hey @Mandipbhattarai ,
The entry LGTM!

Thanks,
Savi

**Comment by Mandipbhattarai** [comment]
@SaviDahegaonkar can i get to know the issue so I can perform the necessary changes if its from my side. 

Thanks,
Mandip

**Comment by Mandipbhattarai** [comment]
@SaviDahegaonkar is there any changes to be done?

**Comment by SaviDahegaonkar** [comment]
> @SaviDahegaonkar is there any changes to be done?

Nope, I will merge this PR.

**Pull request merged**